File: gto-support-api-okhttp3/src/main/java/org/ccci/gto/android/common/api/okhttp3/interceptor/SessionInterceptor.java
package org.ccci.gto.android.common.api.okhttp3.interceptor;

import android.content.Context;
import android.content.SharedPreferences;
import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.ccci.gto.android.common.api.Session;
import org.ccci.gto.android.common.api.okhttp3.EstablishSessionApiException;
import org.ccci.gto.android.common.api.okhttp3.InvalidSessionApiException;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public abstract class SessionInterceptor<S extends Session> implements Interceptor {
    private static final boolean DEFAULT_RETURN_INVALID_SESSION_RESPONSES = false;
    protected final Object mLockSession = new Object();

    @NonNull
    protected final Context mContext;
    private final boolean mReturnInvalidSessionResponses;
    @NonNull
    private final String mPrefFile;

    protected SessionInterceptor(@NonNull final Context context) {
        this(context, DEFAULT_RETURN_INVALID_SESSION_RESPONSES, null);
    }

    protected SessionInterceptor(@NonNull final Context context, final boolean returnInvalidSessionResponses) {
        this(context, returnInvalidSessionResponses, null);
    }

    protected SessionInterceptor(@NonNull final Context context, @Nullable final String prefFile) {
        this(context, DEFAULT_RETURN_INVALID_SESSION_RESPONSES, prefFile);
    }

    protected SessionInterceptor(@NonNull final Context context, final boolean returnInvalidSessionResponses,
                                 @Nullable final String prefFile) {
        mContext = context.getApplicationContext();
        mReturnInvalidSessionResponses = returnInvalidSessionResponses;
        mPrefFile = prefFile != null ? prefFile : getClass().getSimpleName();
    }

    @NonNull
    protected final SharedPreferences getPrefs() {
        return mContext.getSharedPreferences(mPrefFile, Context.MODE_PRIVATE);
    }

    @Override
    public final Response intercept(@NonNull final Chain chain) throws IOException {
        // get the session, establish a session if one doesn't exist or if we have a stale session
        S session;
        synchronized (mLockSession) {
            session = loadSession();
            if (session == null) {
                try {
                    session = establishSession();
                } catch (final IOException e) {
                    // wrap establish session IOExceptions
                    throw new EstablishSessionApiException(e);
                }

                // save the newly established session
                if (session != null && session.isValid()) {
                    saveSession(session);
                }
            }
        }

        // throw an exception if we don't have a valid session
        if (session == null) {
            throw new InvalidSessionApiException();
        }

        // process request & response
        return processResponse(chain.proceed(attachSession(chain.request(), session)), session);
    }

    @NonNull
    protected abstract Request attachSession(@NonNull Request request, @NonNull S session);

    @CallSuper
    protected Response processResponse(@NonNull final Response response, @NonNull S session) throws IOException {
        // make sure the response is valid
        if (isSessionInvalid(response)) {
            // reset the session
            synchronized (mLockSession) {
                // only reset if this is still the same session
                final S active = loadSession();
                if (active != null && active.equals(session)) {
                    deleteSession(session);
                }
            }

            // throw an invalid session exception because our session was invalid
            if (!mReturnInvalidSessionResponses) {
                throw new InvalidSessionApiException();
            }
        }

        return response;
    }

    protected boolean isSessionInvalid(@NonNull final Response response) throws IOException {
        return false;
    }

    @Nullable
    private S loadSession() {
        // load a pre-existing session
        final SharedPreferences prefs = this.getPrefs();
        final S session;
        synchronized (mLockSession) {
            session = loadSession(prefs);
        }

        // only return valid sessions
        return session != null && session.isValid() ? session : null;
    }

    @Nullable
    protected abstract S loadSession(@NonNull SharedPreferences prefs);

    @Nullable
    protected S establishSession() throws IOException {
        return null;
    }

    protected final void saveSession(@NonNull final S session) {
        final SharedPreferences.Editor prefs = this.getPrefs().edit();
        session.save(prefs);

        synchronized (mLockSession) {
            prefs.apply();
        }
    }

    private void deleteSession(@NonNull final S session) {
        final SharedPreferences.Editor prefs = getPrefs().edit();
        session.delete(prefs);

        synchronized (mLockSession) {
            prefs.apply();
        }
    }
}

File: gto-support-api-okhttp3/src/main/java/org/ccci/gto/android/common/api/okhttp3/interceptor/SessionInterceptor.kt
package org.ccci.gto.android.common.api.okhttp3.interceptor

import android.content.Context
import android.content.SharedPreferences
import androidx.annotation.CallSuper
import java.io.IOException
import okhttp3.Interceptor
import okhttp3.Interceptor.Chain
import okhttp3.Request
import okhttp3.Response
import org.ccci.gto.android.common.api.Session
import org.ccci.gto.android.common.api.okhttp3.EstablishSessionApiException
import org.ccci.gto.android.common.api.okhttp3.InvalidSessionApiException

private const val DEFAULT_RETURN_INVALID_SESSION_RESPONSES = false

abstract class SessionInterceptor<S : Session> @JvmOverloads protected constructor(
    context: Context,
    private val returnInvalidSessionResponses: Boolean = DEFAULT_RETURN_INVALID_SESSION_RESPONSES,
    prefFile: String? = null
) : Interceptor {
    protected val context: Context = context.applicationContext
    protected val prefs by lazy {
        this.context.getSharedPreferences(prefFile ?: javaClass.simpleName, Context.MODE_PRIVATE)
    }

    protected val lockSession = Any()

    @JvmField
    @Deprecated("Since v3.10.0, use lockSession property instead", ReplaceWith("lockSession"))
    protected val mLockSession = lockSession
    @JvmField
    @Deprecated("Since v3.10.0, use context property instead", ReplaceWith("context"))
    protected val mContext = this.context

    @Deprecated("Since v3.10.0, use named parameters for the constructor instead.")
    protected constructor(context: Context, prefFile: String?) :
        this(context, DEFAULT_RETURN_INVALID_SESSION_RESPONSES, prefFile)

    @Throws(IOException::class)
    override fun intercept(chain: Chain): Response {
        val session = synchronized(lockSession) {
            // load a valid session
            loadSession()?.takeIf { it.isValid }
                // otherwise try establishing a session
                ?: try {
                    establishSession()?.takeIf { it.isValid }?.also { saveSession(it) }
                } catch (e: IOException) {
                    throw EstablishSessionApiException(e)
                }
        } ?: throw InvalidSessionApiException()

        return processResponse(chain.proceed(attachSession(chain.request(), session)), session)
    }

    protected abstract fun attachSession(request: Request, session: S): Request

    @CallSuper
    @Throws(IOException::class)
    protected open fun processResponse(response: Response, session: S): Response {
        if (isSessionInvalid(response)) {
            // reset the session if this is still the same session
            synchronized(lockSession) {
                loadSession()?.takeIf { it == session }?.apply { deleteSession(this) }
            }

            if (!returnInvalidSessionResponses) throw InvalidSessionApiException()
        }
        return response
    }

    @Throws(IOException::class)
    open fun isSessionInvalid(response: Response) = false

    private fun loadSession() = synchronized(lockSession) { loadSession(prefs) }?.takeIf { it.isValid }
    protected abstract fun loadSession(prefs: SharedPreferences): S?

    @Throws(IOException::class)
    protected open fun establishSession(): S? = null

    private fun saveSession(session: Session) {
        val changes = prefs.edit().apply { session.save(this) }
        synchronized(lockSession) { changes.apply() }
    }

    private fun deleteSession(session: Session) {
        val changes = prefs.edit().apply { session.delete(this) }
        synchronized(lockSession) { changes.apply() }
    }
}